### ISSUE TITLE
Goal site isn't visible for PickCube-v1 when using rgbd observation mode

### ISSUE THREAD

**Issue by jstmn** [haosulab/ManiSkill#1119]
Hi, 

I've noticed that the green goal site isn't visible for PickCube-v1 when in rgbd obs mode. Specifically, the rgb images in the saved demonstrations don't include the site. Is this intended behavior? 

I'm trying to train a neural network to estimate the reward function given an rgb of the scene but am unable to do so because the goal site is missing, which makes the problem under specified. My understanding is that the goal site is provided explicitly when in state based mode, and should be shown in the image when in rgb/pointcloud mode.

![Image](https://github.com/user-attachments/assets/dbf8094a-e14a-44bd-8f1e-7dbf86fcbbf8)

```
uv run python scripts/rgb_reward_function/collect_demos.py \
    --env-id PickCube-v1 \
    --num-traj 1 \
    --only-count-success \
    --reward-mode dense \
    --obs-mode rgbd \
    --sim-backend cpu \
    --traj-name traj \
    --record-dir /tmp/msbug/ \
    --num-procs 1 \
    --render-mode rgb_array
```
next, in a python script:
```
import h5py
from PIL import Image

with h5py.File("/tmp/msbug/PickCube-v1/motionplanning/traj.h5", "r") as f:
    rgbs = f["traj_0"]["obs"]["sensor_data"]["base_camera"]["rgb"]
    Image.fromarray(rgbs[0]).save("/tmp/msbug/PickCube-v1/motionplanning/traj_0-step_0.png")
```

Thanks

**Comment by chenyenru** [comment]
I think that's intentional. Goal site was added to hidden objects to be hidden from the sensor: https://github.com/haosulab/ManiSkill/blob/main/mani_skill/envs/tasks/tabletop/pick_cube.py#L105

If you want to unhide you can remove the appending

**Comment by jstmn** [comment]
Oh, thanks I see. Not to be rude but I feel there is some room for improvement in how this is done. Especially because objects in this hidden list are still visible in the Gui.

At a minimum, throwing a warning that the objects in the `_hidden_objects` list aren't visible would be helpful/informative.

**Comment by jstmn** [comment]
```
python scripts/rgb_reward_function/collect_demos.py \
    --env-id PickCube-v1 \
...
    --obs-mode rgbd \
```
like it seems wrong that this command doesn't generate 'complete' demonstrations ^

there should be a loud warning or an error thrown imo!

**Comment by StoneT2000** [comment]
It should. For these tasks the goal position is in the extra key of the observation sequence.


This is a bit of a legacy thing we kept from MS2 where the goal site isn't visible and we specify the goal via 3d position

I'm not sure what's the best kind of warning to give?

**Comment by jstmn** [comment]
For a first pass, could add a simple printout in `BaseEnv.__init__()`:
```python
for obj in self._hidden_objects:
            print(f"WARNING: object {obj.name} is hidden")
```
making the printout yellow with `cprint()` from [termcolor](https://pypi.org/project/termcolor/) could be good too.

Not sure about a good long term solution, i'd need to think about this more

Happy to submit a PR with this, lmk

**Comment by jstmn** [comment]
also, a related question - in pointcloud mode the goal region isn't included, so the only way to know where the goal is is from the key in the observation sequence right? Which means for policy learning that means your model should accept the pointcloud and also the target pose?

**Comment by StoneT2000** [comment]
I'm still not sure about the hidden warning. 

But maybe it depends on who you are. If you are making custom envs you absolutely want to know this.

If you are just doing ML you probably don't. You just use the entire obs dictionary as you see fit.

There are also some envs outside of this repo that like to hide and unhide objects during run time. 

As a result I think it is best not to print such a warning like that. 

Re the question about the pointcloud. The same answer applies, for all observation modes, the expectation is that you use every key / value in the returned observation data in order to solve the task. all our IL baselines do this and read every datapoint.

We generally ensure all tasks are solvable if you use all datapoints.


The only ones that might not be are like plug charger or assembly kits with visual obs modes but that is only because they are way too hard to solve with end to end IL/RL at the moment.

**Comment by jstmn** [comment]
All good points. I agree that fundamentally, the problem is poorly or non specified if using rgbd/pointcloud observations without the goal site pose visible. 

Looking through some example code, it seems like the reason this is happening is that i've been using `model(obs["sensor_data"]["camera_center"]["rgb"])` rather than `model(common.to_tensor(obs, device))`

I still think some kind of printout would be good to include, but this isn't my hill to die on

**Comment by jstmn** [comment]
Wait, i'm sure i'm missing something obvious here but how do I convert `obs` to a single tensor? Is that something I need to write myself?

neither `common.to_tensor()` nor `common.flatten_state_dict()` is doing what I expect. Here's what my `obs` looks like:

```
obs:
    - agent: dict
        - qpos: Tensor torch.Size([84, 9]) torch.float32
        - qvel: Tensor torch.Size([84, 9]) torch.float32
    - extra: dict
        - is_grasped: Tensor torch.Size([84]) torch.bool
        - tcp_pose: Tensor torch.Size([84, 7]) torch.float32
        - goal_pos: Tensor torch.Size([84, 3]) torch.float32
    - sensor_param: dict
        - camera_center: dict
            - extrinsic_cv: Tensor torch.Size([84, 3, 4]) torch.float32
            - cam2world_gl: Tensor torch.Size([84, 4, 4]) torch.float32
            - intrinsic_cv: Tensor torch.Size([84, 3, 3]) torch.float32
    - sensor_data: dict
        - camera_center: dict
            - rgb: Tensor torch.Size([84, 240, 320, 3]) torch.uint8
            - depth: Tensor torch.Size([84, 240, 320, 1]) torch.int16
```

**Comment by StoneT2000** [comment]
We usually use the FlattenRGBD wrapper (see ppo code for example). You can ditch the sensor parameters.

You can flatten state data, and combine all images together, to make a dict obs with state and rgb key. 

You generally shouldn't flatten image data.

**Comment by jstmn** [comment]
I see, thanks. To make sure I have this right you'll want to do 

model(obs["state"], obs["rgbd"])
```
obs, FlattenRGBDObservationWrapper:
    - state: Tensor torch.Size([84, 29]) torch.float32
    - rgbd: Tensor torch.Size([84, 240, 320, 4]) torch.int16
```

**Comment by StoneT2000** [comment]
Yes, I would recommend this way of doing it. I recommend decoupling the RGB and depth image data since the depth data is also scaled (it is returned raw in millimeters by default).

**Comment by jstmn** [comment]
OK nice. Thanks for the pointer.
